Pay some technical debt

matte21 opened the pull request:
**Merge #76 first**

This PR was rebased on the pulled branch from #76, it's WIP because #76 has not been merged yet.
 
**What does this PR do**

- Tweak Grafana dashboards for IPAM and Subnet Validator as described [here](https://github.com/MikeSpreitzer/kube-examples/pull/75#issuecomment-514646405): merge dashboards about worker utilization and worker utilization spent on calls out. Tweaks and additions concerning the Connection Agent are left for after the refactoring of said controller. 
- Add explanatory comment in subnet strategy code.

matte21 committed:
Pay some technical debt

Merge Grafana dashboards on worker utilization and worker utilization
spent on calls out because they are related and it's useful to see them
in comparison. Do that for IPAM Controller and Subnet Validator only
because the story is more complex with the Connection Agent and because
it will go through a heavy overhaul soon.

Add comment to subnet strategy code on why existing subnets are checked
to make sure they are well formed. This is far from obvious and needs
explanation.

matte21 commented on staging/kos/pkg/registry/network/subnet/strategy.go:
```diff
@@ -137,6 +137,10 @@ func (ss *subnetStrategy) checkNSAndCIDRConflicts(candidate *subnet.Summary) (er
+		// Make sure potentialRival is well formed. The code in this file makes
+		// it impossible to create a malformed subnet, but the check is done in
+		// case this version of strategy is rolled out after a more lenient one
+		// which let malformed subnets through.
```
See https://github.com/MikeSpreitzer/kube-examples/pull/69#discussion_r302189214. Your final recommendation was to remove the check below the added comment because we're not in production and because it's far from obvious why the check might be desirable. I propose to simply put an explanatory comment. If you still want the check removed I'll do that.

MikeSpreitzer reviewed (commented):
Some comments on the kos-internals Grafana dashboard.

The "number of workers" and "number of CA, by fabric name" panels are the least interesting, they just confirm configuration.  It is odd to me that they get full width while much more interesting panels, like the ones for work queue stats, get only half width.  I would prefer it the other way around.

Some utilization panels have traces named "total work", but "work" has interesting dimensions while "utilization" is dimensionless.  I think it would suffice to name those traces "total" rather than "total work".

The common pattern for worker utilization is to exhibit utilization, 99th percentile of latency for each kind of work , average latency for each kind of work, and rate of each kind of work.  That is mostly followed.  But rate of status updates from CAs is missing; I think it was there earlier.  Also the rates of fabric calls are missing; I think they were never there, and have not particularly missed them.

I am surprised that the IPAM worker utilization is not almost entirely due to calls out; I thought it was in earlier versions.

<img width="567" alt="Screen Shot 2019-08-16 at 12 00 55 AM" src="https://user-images.githubusercontent.com/14296719/63142767-20b37e80-bfb9-11e9-9fa5-eeb971ccaae5.png">

matte21 committed:
Tweak Grafana Dashboards as suggested in review

matte21 commented:
> The "number of workers" and "number of CA, by fabric name" panels are the least interesting, they just confirm configuration. It is odd to me that they get full width while much more interesting panels, like the ones for work queue stats, get only half width. I would prefer it the other way around.

Fixed this in b222442. Notice how "number of CA, by fabric name" panel moves to the top at the side of "number of workers". This breaks the pattern of having all the controller-wise metrics at the top and the per-controller metrics below and grouped per-controller. I think that's ok.

> Some utilization panels have traces named "total work", but "work" has interesting dimensions while "utilization" is dimensionless. I think it would suffice to name those traces "total" rather than "total work".

Fixed in b222442.

> The common pattern for worker utilization is to exhibit utilization, 99th percentile of latency for each kind of work , average latency for each kind of work, and rate of each kind of work. That is mostly followed. But rate of status updates from CAs is missing; I think it was there earlier. Also the rates of fabric calls are missing; I think they were never there, and have not particularly missed them.

Yes, the CA metrics miss some stuff. I plan to revamp them right after the PR simplifying the CA (which is in progress).

> I am surprised that the IPAM worker utilization is not almost entirely due to calls out; I thought it was in earlier versions.

I don't recall much from the past experiments and earlier versions performance, but I agree, it's counterintuitive. The IPAM doesn't do anything too computationally expensive between API calls.  
That might be something to look at more in depth during the performance study.
